File: src/main/java/org/gd/leetcode/p0112/Solution.java
package org.gd.leetcode.p0112;

import org.gd.leetcode.common.LeetCode;
import org.gd.leetcode.common.TreeNode;

/**
 * https://leetcode.com/problems/path-sum/
 *
 * @see org.gd.leetcode.p0113.Solution
 * @see org.gd.leetcode.p0257.Solution
 */
@LeetCode(
        difficulty = LeetCode.Level.EASY,
        tags = {
                LeetCode.Tags.TREE,
                LeetCode.Tags.DEPTH_FIRST_SEARCH
        }
)
class Solution {

    private static boolean traverse(TreeNode root, int sum, int target) {
        sum += root.val;
        if (root.left == null && root.right == null)
            return sum == target;

        return (root.left != null && traverse(root.left, sum, target))
                || (root.right != null && traverse(root.right, sum, target));
    }

    public boolean hasPathSum(TreeNode root, int sum) {
        return root != null && traverse(root, 0, sum);
    }
}

File: src/main/java/org/gd/leetcode/p0112/solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def hasPathSum(self, root: TreeNode, target_sum: int) -> bool:

        def traverse(node: TreeNode, path_sum: int) -> bool:
            path_sum += node.val
            if not node.left and not node.right:
                return target_sum == path_sum

            return (node.left and traverse(node.left, path_sum)) or (node.right and traverse(node.right, path_sum))

        return root and traverse(root, 0)
